spring-cloud-aws-core/src/main/java/org/springframework/cloud/aws/core/io/s3/AmazonS3ProxyFactory.java

```diff
@@ -131,11 +132,34 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
 			}
 		}
 
+		/**
+		 * Builds a new S3 client based on the information from the
+		 * {@link AmazonS3Exception}. Extracts from the exception's additional details the
+		 * region and endpoint of the bucket to be redirected to.
+		 *
+		 * Extracting the region from the exception is needed because the US S3 buckets
+		 * don't always return an endpoint that includes the region and
+		 * {@link AmazonS3ClientFactory} will default to us-west-2 if the hostname of
+		 * the endpoint is "s3.amazonaws.com". The us-east-1 bucket is quite likely to
+		 * return the "s3.amazonaws.com" endpoint.
+		 *
+		 * @author André Caron
+		 */
 		private AmazonS3 buildAmazonS3ForRedirectLocation(AmazonS3 prototype,
 				AmazonS3Exception e) {
 			try {
+				Regions redirectRegion;
+				try {
+					redirectRegion = Regions.fromName(
+							e.getAdditionalDetails().get("x-amz-bucket-region"));
+				}
+				catch (IllegalArgumentException iae) {
+					redirectRegion = null;
+				}
+
 				return this.amazonS3ClientFactory.createClientForEndpointUrl(prototype,
-						"https://" + e.getAdditionalDetails().get("Endpoint"));
+						"https://" + e.getAdditionalDetails().get("Endpoint"),
```

Comment:
As far as I can test it with real S3 scenarios, this is always `null`. Meaning, PR does not completely fix the problem as getting 301 still ends up with exception. Do we actually need to pass this endpoint there? Isn't region enough to create client?

Reply:
It's strange, from my tests, whenever I had the redirect, I did get the endpoint value. I kept the endpoint part as close as possible to the current behavior. It's also required according to `createClientForEndpointUrl`;
https://github.com/spring-cloud/spring-cloud-aws/pull/502/files#diff-3afe91a30772bd065279abd0e4d3aee9R102

Can I ask you in which regions you did your tests, so I can test them as well? I tried with US and EU buckets. `us-east-1` behaves differently then the other US bucket regions. Maybe this region you tried may also be different. It's also been a while so it's possible the behavior has changed.

Reply:
I had client set to eu-west-1 and bucket in eu-central-1. If you still won't be able to reproduce I'll put on GitHub sample application that reproduces the issue (but it won't happen today)

Reply:
I'll need a few days as well. But I'll try with those buckets and I'll publish my results here.

Reply:
I did several tests with buckets from different regions using clients in another region form the bucket. I got consistent behavior for all redirection cases. I never received a `null` value from either `Endpoint` or `x-amz-bucket-region`. 


For the test I just added the following code in the `createClientForEndpointUrl_withDefaultRegionUrl_createClientForDefaultRegion` after line 77
```
		AmazonS3 proxy = AmazonS3ProxyFactory.createProxy(amazonS3);

		GetObjectRequest request = new GetObjectRequest("<bucket name here>","text.txt");
		S3Object ob = proxy.getObject(request);
```

I played around with the region at line 77 of the test and used buckets from different regions. While debugging I then put a breakpoint at line 153 of `AmazonS3ProxyFactory` to see the exception's additional details. Here's a list of some of the use cases I tired and their values

```
Regions.AP_NORTHEAST_1 => cseg-test-spring-useastone
"Endpoint" -> "s3.amazonaws.com"
"Bucket" -> "cseg-test-spring-useastone"
"Error" -> "Q4C9Ipoij1AC1sL+jVzoHeWuSSCFLnhE6lYQTi0fjv84KdZJsmku0aZMcy/ZnIzzR8PMe0ncgu0="
"x-amz-bucket-region" -> "us-east-1"

Regions.EU_WEST_1 => cseg-test-spring-useastone
"Endpoint" -> "s3.amazonaws.com"
"Bucket" -> "cseg-test-spring-useastone"
"Error" -> "dReOcKIHBudIBAEpgRavx14A6sQ13ZRdrwh07ncrjhlmTHu9I8eYpUbfin8rHmgGcXfO9uMhGIs="
"x-amz-bucket-region" -> "us-east-1"

Regions.US_WEST_2 => cseg-test-spring-useastone
"Endpoint" -> "s3.amazonaws.com"
"Bucket" -> "cseg-test-spring-useastone"
"Error" -> "gdkyBuzJ4miDHvP7lbVXSxSgWc9CWVMDNpmkV712ET1ug8f/Ds0gakV3eh8nf8k/hPIDfKgPnuU="
"x-amz-bucket-region" -> "us-east-1"

Regions.EU_WEST_1 => cseg-test-spring-eucentralone
"Endpoint" -> "cseg-test-spring-eucentralone.s3.eu-central-1.amazonaws.com"
"Bucket" -> "cseg-test-spring-eucentralone"
"Error" -> "JtWtlKQXOEd8aahSFuF67UU3hgZcZVMj77vbwUSy29kZz1V93fonQh3bS59S/0vKgMoDYfn4cAA="
"x-amz-bucket-region" -> "eu-central-1"

Regions.AP_NORTHEAST_1 => cseg-test-spring-uswesttwo
"Endpoint" -> "cseg-test-spring-uswesttwo.s3-us-west-2.amazonaws.com"
"Bucket" -> "cseg-test-spring-uswesttwo"
"Error" -> "RJDShD97Vo2xVxCi0EX0ZqJWOZQLPyUE9T1cJgEYNoyRyFn35L3FieGO/e60U8i2qWl4xqZDgoo="
"x-amz-bucket-region" -> "us-west-2"

Regions.US_WEST_2 => cseg-test-spring-eucentralone
"Endpoint" -> "cseg-test-spring-eucentralone.s3.eu-central-1.amazonaws.com"
"Bucket" -> "cseg-test-spring-eucentralone"
"Error" -> "GZrL2EUeq+3Vi72nxejFc94dg2YTOrQEaaqQzrWzcWGLwK36exsSrS4X5vaSggYei23zo5HCiSc="
"x-amz-bucket-region" -> "eu-central-1"
```

Reply:
I'll try to reproduce the issue I was facing and post code on GitHub. Thanks for investigation!

Reply:
I reproduced this issue here: https://github.com/maciejwalkowiak/spring-cloud-aws-gh-502/blob/master/src/main/java/com/example/gh502test/Gh502TestApplication.java

Take a look please and let me know if anything is unclear.

Reply:
Thanks @maciejwalkowiak. I cloned your project and I'll experiment with it against my buckets. Again, it may take me a few days to get to it.

Reply:
Hello @maciejwalkowiak, sorry for taking so long, but I did manage to find some time to look into the issue. I tried your project as is, and got the null `Endpoint`. 

I didn't yet look into why sometimes we have `Endpoint` and sometimes not, but I did try your project with a SNAPSHOT of my branch. The changes I propose do fix the issue since the exception from AWS does at least include `x-amz-bucket-region`. 

My proposal leaves the current behavior as much as possible as it is today. However, when the 301 exception is returned, it favors the region from `x-amz-bucket-region` over the one extracted from the `Endpoint` value (when it's available). Is there still a change you'd like me to do here?

Reply:
Hello! Thanks for looking into. Yes, please this sounds like a good approach!

Reply:
That's great! That's how it is :) Is there a next step? 

I noticed the classes I touched don't quite follow the code style described in the README. I followed how it looked in the class now, but should I apply the style to the whole class? Or just the parts I changed?

Reply:
Hello @maciejwalkowiak, I forgot to mention you when I wrote my last comment ⬆️. Let me know if there is anything else needed.

Reply:
No worries I get all the notifications. Leave this PR as it is - we are merging it, we just have some internal discussions how to merge changes between branches as we figured out we followed different approaches.

Reply:
That's great! Don't hesitate to reach out if there are any more questions. Good luck merging the branches!